spacy/kb/candidate.pyx

```diff
@@ -1,74 +1,114 @@
 # cython: infer_types=True, profile=True
 
-from typing import Iterable
-from .kb cimport KnowledgeBase
-from ..tokens import Span
+from .kb_in_memory cimport InMemoryLookupKB
+from ..errors import Errors
 
 cdef class Candidate:
-    """A `Candidate` object refers to a textual mention (`alias`) that may or may not be resolved
-    to a specific `entity` from a Knowledge Base. This will be used as input for the entity linking
+    """A `Candidate` object refers to a textual mention that may or may not be resolved
+    to a specific entity from a Knowledge Base. This will be used as input for the entity linking
     algorithm which will disambiguate the various candidates to the correct one.
-    Each candidate (alias, entity) pair is assigned a certain prior probability.
+    Each candidate (mention, entity_id) pair is assigned a certain prior probability.
 
     DOCS: https://spacy.io/api/kb/#candidate-init
     """
 
-    def __init__(self, KnowledgeBase kb, entity_hash, entity_freq, entity_vector, alias_hash, prior_prob):
-        self.kb = kb
-        self.entity_hash = entity_hash
-        self.entity_freq = entity_freq
-        self.entity_vector = entity_vector
-        self.alias_hash = alias_hash
-        self.prior_prob = prior_prob
+    def __init__(
+        self,
+        entity_id: str,
+        entity_vector: vector[float],
+        prior_prob: float,
+    ):
+        """Initializes properties of abstract base class `Candidate`.
+        entity_id (Union[str, int]): Unique entity ID.
+        entity_vector (List[float]): Entity embedding.
+        prior_prob (float): Prior probability of entity for this mention - i.e. the probability that, independent of
+            the context, this mention resolves to this entity_id in the corpus used to build the knowledge base. In
+            cases in which this isn't always possible (e.g.: the corpus to analyse contains mentions that the KB corpus
+            doesn't) it might be better to eschew this information and always supply the same value.
```

Comment:
Regarding the mention/alias terminology: this is a bit tricky with `prior_prob`. It only really makes sense for _aliases_, since we don't and realistically can't have information on the prior probabilities of arbitrary mentions resolving to specific entities, just for the specific aliases stored in the KB.

The context in which this is used though is a "mention". As in, for "Boston play tonight." candidates for "Boston" would refer to a _mention_, not an alias. This points to a deeper problem: 
- Prior probabilities only make sense if we are operating if mention == alias. This won't always be the case. How do we cope with situations in which some mentions have exactly matching aliases, and therefore a non-zero prior, and others don't? Will the former always be preferred then?
E. g.: "Boston" may refer to the basketball team "Boston Celtics", but ["Boston" isn't listed as an alias](https://www.wikidata.org/wiki/Q131371). If we accept that the city of Boston has a non-zero prior (as "Boston" would certainly be a listed alias), this automatically puts "Boston Celtics" at an disadvantage. This doesn't seem great.
- The concept of prior probability for candidate retrieval without exact lexical matching seems questionable for me due to that. Should this be an implementation detail of `InMemoryLookupKB` (and `InMemoryCandidate`)?
- If it should be an implementation detail - how should `EntityLinker` deal with that? Assume `prior_prob` is zero if it's not available in the candidate instance?